### PR TITLE
Add linter for checking blog posts

### LINKED ISSUE #2432: Add linter for checking blog posts
### What should be done?

Our [blog](https://blog.ferretdb.io) uses the Docusaurus engine. All blog posts are stored in the `website/blog` directory as Markdown files. In addition to existing Markdown linters and Docusaurus checks, we want to introduce a new linter to check a few things about our blogging process.

Let's start by checking that all files in the `website/blog` directory start with some data in `YYYY-MM-DD` format, followed by `-`, "slug", and `.md` extension. The "slug" is a part of the URL path for that blog post. That same slug also should be present in the file itself in the frontmatter section.

Let's create a new linter in `tools/checkdocs` directory. It should be written in Go, not use external dependencies, not try to parse Markdown "properly" (regular expressions with `bufio.Scanner` will do just fine), and check filenames and slugs as written above. That linter should run in `docs-fmt` Taskfile target.

### Where?

`tools/checkdocs` directory.

### Definition of Done

Linter exists and works.

### PR DESCRIPTION
**Pull request by raeidish**
# Description

Closes #2432.

Added linter in tools/checkdocs that verifies slugs in markdown files. Would love to get some feedback on the changes 

## Readiness checklist

* [x] I added/updated unit tests.
* [ ] I added/updated integration/compatibility tests.
* [ ] I added/updated comments and checked rendering.
* [ ] I made spot refactorings.
* [ ] I updated user documentation.
* [x] I ran `task all`, and it passed.
* [ ] I ensured that PR title is good enough for the changelog.
* [ ] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

**Comment by AlekSi** [comment]
> Would love to get some feedback on the changes

Then please mark this PR as ready for review

File: tools/Taskfile.yml [hunk 1]
```diff
@@ -9,6 +9,7 @@ tasks:
   tools-test:
     cmds:
       - go test -short {{.RACEFLAG}} -shuffle=on -coverprofile=cover.txt -coverpkg=./... ./checkswitch/...
+      - go test ./checkdocs/
```
Comment (AlekSi) [inline tools/Taskfile.yml:12]:
```suggestion
      - go test -short {{.RACEFLAG}} -shuffle=on -coverprofile=cover.txt -coverpkg=./... ./checkdocs/... ./checkswitch/...
```

File: Taskfile.yml [hunk 2]
```diff
@@ -551,7 +551,9 @@ tasks:
   # see https://github.com/DavidAnson/markdownlint-cli2#command-line for the reason we use double-quotes
   docs-fmt:
     desc: "Format and lint documentation"
+    deps: [init-tools]
     cmds:
+      - bin/checkdocs -i "website/blog"
```
Comment (AlekSi) [inline Taskfile.yml:556]:
```suggestion
      - bin/checkdocs
```

File: tools/checkdocs/checkdocs.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	"regexp"
+)
+
+type FileSlug struct {
+	slug     string
+	fileName string
+}
+
+func main() {
+	dir := flag.String("i", "website/blog", "root of website blog directory")
```
Comment (AlekSi) [inline tools/checkdocs/checkdocs.go:20]: Let's drop that flag and always process `filepath.Join("website", "blog")`

File: tools/checkdocs/checkdocs_test.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"testing"
+	"testing/fstest"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetBlogSlugs(t *testing.T) {
+	m := fstest.MapFS{
```
Comment (AlekSi) [inline tools/checkdocs/checkdocs_test.go:11]: 👍

File: tools/checkdocs/checkdocs.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		sm := r.FindStringSubmatch(scanner.Text())
+		if len(sm) > 1 && sm[len(sm)-1] == fS.slug {
+			pass = true
+		}
+	}
+
+	if !pass {
+		log.Fatalf("Slug is not correctly formated in file: %s", fS.fileName)
```
Comment (AlekSi) [inline tools/checkdocs/checkdocs.go:83]: Let's return an error from that function, and make the caller to print all errors before exiting with non-zero exit code

File: tools/checkdocs/checkdocs.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	dir := flag.String("i", "website/blog", "root of website blog directory")
+
+	fs, err := os.ReadDir(*dir)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	slugs := GetBlogSlugs(fs)
+	for _, slug := range slugs {
+		fo, err := os.Open(filepath.Join(*dir, slug.fileName))
```
Comment (chilagrow) [inline tools/checkdocs/checkdocs.go:29]:
```suggestion
		fo, err := os.Open(filepath.Join(*dir, slug.fileName))
		defer fo.Close()
```

Let's make sure to close the file, defer closes it when this function returns.

**Comment by raeidish** [comment]
Ready for another review @AlekSi

**Comment by AlekSi** [comment]
CleanShot 2023-04-19 at 13 58 56@2x

Please click those

File: tools/checkdocs/checkdocs_test.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+image: /img/blog/cla3.jpg
+date: 2022-05-16
+---
+Finally, we need a web server that would handle HTTPS for us.
+For that, we will use [Caddy](https://caddyserver.com):
+Caddy will listen on both HTTP and HTTPS ports, and retrieve the TLS certificate from Let’s Encrypt that will be stored in “./data/caddy” on the host.
+`),
+		},
+	}
+	dirs, _ := m.ReadDir(".")
```
Comment (AlekSi) [inline tools/checkdocs/checkdocs_test.go:49]: Let's import `github.com/stretchr/testify/require` and

```suggestion
	dirs, err := m.ReadDir(".")
	require.NoError(t, err)
```

File: Taskfile.yml [hunk 8]
```diff
@@ -558,7 +558,9 @@ tasks:
   # see https://github.com/DavidAnson/markdownlint-cli2#command-line for the reason we use double-quotes
   docs-fmt:
     desc: "Format and lint documentation"
+    deps: [init-tools]
```
Comment (AlekSi) [inline Taskfile.yml:561]: We expect people to run `task init` anyway, and that dependency slows things down. Let's remove it

File: tools/checkdocs/checkdocs.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+	"fmt"
+	"io"
+	"io/fs"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+)
+
+type FileSlug struct {
```
Comment (AlekSi) [inline tools/checkdocs/checkdocs.go:14]: Let's add comments to all top-level types and function (except `main`)

File: tools/checkdocs/checkdocs.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+	slug     string
+	fileName string
+}
+
+func main() {
+    dir := filepath.Join("website","blog")
```
Comment (AlekSi) [inline tools/checkdocs/checkdocs.go:20]: Please pull my commits and run `task fmt lint`

File: tools/checkdocs/checkdocs.go [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+	fs, err := os.ReadDir(dir)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	slugs := GetBlogSlugs(fs)
+    pass := true
+	for _, slug := range slugs {
+
```
Comment (AlekSi) [inline tools/checkdocs/checkdocs.go:29]: Let's remove that blank line

**Review by chilagrow — COMMENTED** [review]
Very minor comments. Looking good :hugs:

**Review by chilagrow — COMMENTED** [review]
Could you run `task lint` and fix the warnings of that please? :hugs:

**Comment by raeidish** [comment]
> Could you run `task lint` and fix the warnings of that please? hugs

@chilagrow  what linting errors are you getting? because apparently I'm not getting any. 

works on my machine :grin:

**Comment by chilagrow** [comment]
> > Could you run `task lint` and fix the warnings of that please? hugs
> 
> @chilagrow what linting errors are you getting? because apparently I'm not getting any.
> 
> works on my machine grin

That's interesting, below is what I see from `task lint` @raeidish :thinking: 
<details>
  <summary>task lint</summary>

```
task lint
task: [lint] bin/golangci-lint run --config=.golangci.yml
task: [lint] bin/golangci-lint run --config=.golangci-new.yml
task: [lint] bin/go-consistent -pedantic ./cmd/... ./internal/... ./build/... ./ferretdb/...
task: [lint] bin/go-sumtype ./...
task: [lint] go vet -vettool=./bin/checkswitch ./...
task: [lint] bin/task -d integration integration-lint
task: [integration-lint] ../bin/golangci-lint run --config=.golangci.yml
task: [integration-lint] ../bin/golangci-lint run --config=.golangci-new.yml
task: [integration-lint] ../bin/go-consistent -pedantic ./...
task: [integration-lint] ../bin/go-sumtype ./...
task: [integration-lint] go vet -vettool=../bin/checkswitch ./...
task: [lint] bin/task -d tools tools-lint
task: [tools-lint] ../bin/golangci-lint run --config=../.golangci.yml ./checkdocs/... ./checkswitch/...
task: [tools-lint] ../bin/golangci-lint run --config=../.golangci-new.yml ./checkdocs/... ./checkswitch/...
Error: checkdocs/checkdocs.go:15:1: package-comments: should have a package comment (revive)
package main
task: Failed to run task "tools-lint": exit status 1
task: Failed to run task "tools-lint": exit status 1
task: Failed to run task "lint": exit status 201
task: Failed to run task "lint": exit status 201
^
Error: checkdocs/checkdocs.go:37:2: only one cuddle assignment allowed before if statement (wsl)
	if err != nil {
	^
Error: checkdocs/checkdocs.go:43:2: only one cuddle assignment allowed before range statement (wsl)
	for _, slug := range slugs {
	^
Error: checkdocs/checkdocs.go:72:2: only one cuddle assignment allowed before range statement (wsl)
	for _, f := range fs {
	^
Error: Process completed with exit code 201.
```

</details>

**Review by chilagrow — APPROVED** [review]
Great work! This is already helping us to identify file name and slug mismatch.

**Review by w84thesun — APPROVED** [review]
LGTM! Thanks for your contribution! 👍

**Pull request merged**